build.sbt refactor + plugin for easy installation

kondaurovDev opened the pull request:
Hello!
I've added sbt plugin for this library for easy usage. What do you think?

kondaurovDev committed:
build.sbt refactor + plugin for easy installation

kondaurovDev commented on build.sbt:
```diff
@@ -10,37 +10,31 @@ lazy val runtime = (project in file("runtime"))
     name := "GrpcGatewayRuntime",
     libraryDependencies ++= Seq(
-      "com.trueaccord.scalapb" %% "compilerplugin"          % "0.6.7",
-      "com.trueaccord.scalapb" %% "scalapb-runtime-grpc"    % "0.6.7",
```
that was written twice

kondaurovDev commented on build.sbt:
```diff
@@ -10,37 +10,31 @@ lazy val runtime = (project in file("runtime"))
       "org.webjars"            %  "swagger-ui"              % "3.5.0",
-      "com.google.api.grpc"    % "googleapis-common-protos" % "0.0.3" % "protobuf"
-    ),
-    PB.protoSources in Compile += target.value / "protobuf_external",
```
We don't need these settings, do we?

btlines replied:
This is where the `http` option is defined

kondaurovDev replied:
Yes, but everything works without that dependency. runtime doesn't need protos, they are generated at compile time

kondaurovDev commented on build.sbt:
```diff
@@ -10,37 +10,31 @@ lazy val runtime = (project in file("runtime"))
     libraryDependencies ++= Seq(
       "com.trueaccord.scalapb" %% "compilerplugin"          % "0.6.7",
-      "com.trueaccord.scalapb" %% "scalapb-runtime-grpc"    % "0.6.7",
-      "com.google.api.grpc"    % "googleapis-common-protos" % "0.0.3" % "protobuf"
```
why do we depend on scalapb-runtime-grpc? We don't need this, because it's just source code generator

btlines replied:
In theory it's not needed but I think there is a compile error if missing

kondaurovDev replied:
I believe user has to specify this dependency "googleapis-common-protos" explicitly, no need to depend on this in generator. I don't have any compile errors.
"generator" artifact has to depend only on libraries that it needs. There's only one dependency "compilerplugin"

kondaurovDev committed:
for travis

kondaurovDev committed:
Merge remote-tracking branch 'upstream/master' into plugin

# Conflicts:
#	build.sbt

kondaurovDev committed:
in progress

kondaurovDev commented:
I've added example project that uses that library but it's not ready yet. I can't build gateway module, i will look at it later. What do you think about plugin? It it good solution?

btlines commented:
Thanks for the PR!
The plugin approach certainly makes things easier on the user side.
I leave this open for a few days to see what people think ... and I'll get back to it after the holidays

kondaurovDev committed:
Merge remote-tracking branch 'upstream/master' into plugin

# Conflicts:
#	build.sbt

kondaurovDev committed:
in progress

kondaurovDev committed:
in progress

kondaurovDev committed:
in progress

kondaurovDev committed:
in progress

kondaurovDev committed:
in progress

kondaurovDev committed:
stable

kondaurovDev commented:
I've added example project, don't worry that there are many new files :) I've refactored gateway generator and swagger generator a little bit. I think that's all for now that i want to add in this PR

btlines commented on examples/first/.gitignore:
```diff
@@ -0,0 +1,3 @@
+target
+.idea
```
I wouldn't put anything related to a specific IDE in .gitignore

kondaurovDev replied:
Yep, i forgot it, i've created this commit from another machine where global ignore isn't configured :)

btlines commented on examples/first/.gitignore:
```diff
@@ -0,0 +1,3 @@
+target
+.idea
+**/specs/*.yml
```
Can you end your files with an empty line?

kondaurovDev replied:
I will search for this option in idea intellij

btlines commented on examples/first/build.sbt:
```diff
@@ -0,0 +1,67 @@
+scalaVersion in ThisBuild := "2.12.2"
```
We can even go with 2.12.4

kondaurovDev replied:
Ok, i will fix it

btlines commented on examples/first/build.sbt:
```diff
@@ -0,0 +1,67 @@
+
+resolvers in ThisBuild += Resolver.bintrayRepo("beyondthelines", "maven")
+
+disablePlugins(RevolverPlugin)
```
Not sure about using the RevolverPlugin

kondaurovDev replied:
This is example project and moreover RevolverPlugin add ability to run multiple projects in forked process. We can start service and gateway separately and link them via port, do you know other solutions? I think it's like docker's goal :)

btlines commented on generator/src/main/scala/grpcgateway/generators/GatewayGenerator.scala:
```diff
@@ -5,47 +5,25 @@ import com.google.api.HttpRule.PatternCase
+    val b = CodeGeneratorResponse.File.newBuilder()
+    val packageName: List[String] = serviceDescriptor.getFile.scalaPackagePartsAsSymbols.toList
+    b.setName(s"${packageName.mkString("/")}/${serviceDescriptor.getName}Handler.scala")
+    if (params.flatPackage) {}
```
This line can be removed

kondaurovDev replied:
ok, i'll fix it

kondaurovDev committed:
ref

kondaurovDev committed:
up sbt and scala

kondaurovDev committed:
Merge branch 'plugin' of github.com:kondaurovDev/grpcgateway into plugin

# Conflicts:
#	examples/first/build.sbt

kondaurovDev committed:
Merge remote-tracking branch 'upstream/master' into plugin

# Conflicts:
#	build.sbt
#	generator/src/main/scala/grpcgateway/generators/GatewayGenerator.scala
#	generator/src/main/scala/grpcgateway/generators/SwaggerGenerator.scala

kondaurovDev committed:
Merge branch 'plugin' of github.com:kondaurovDev/grpcgateway into plugin

# Conflicts:
#	examples/first/build.sbt
